### ISSUE TITLE
storage: Support parallel composite uploads for faster transfers

### ISSUE THREAD

**Issue by stanhu** [googleapis/google-cloud-go#3219]
**Is your feature request related to a problem? Please describe.**

`gsutil` supports [parallel composite uploads](https://cloud.google.com/storage/docs/uploads-downloads#parallel-composite-uploads), but the [Google Cloud Storage SDKs do not](https://cloud.google.com/storage/docs/uploads-downloads#go). 

For large files, transfers to Google Cloud Storage could be significantly faster if chunks of data are updated simultaneously. The Amazon SDKs natively support multi-threaded multipart uploads with Golang, Ruby, and Python libraries.

**Describe the solution you'd like**

[gsutil implements this in Python](https://github.com/GoogleCloudPlatform/gsutil/blob/4c634ad85dbd13c3445e2e96320c3d91db9ca73e/gslib/utils/copy_helper.py#L1055-L1278).

Ideally there would be a simple boolean parameter, such as `ParallelCompositeUpload`, that would activate this mode in the bucket `Write`.

**Describe alternatives you've considered**

This could be implemented on the application layer, but it really belongs in the SDK.

**Additional context**

**Comment by tritone** [comment]
I'd say this is comparable to #1672. It's a very reasonable feature request, but would entail more involved client-side logic than we usually include in this library, which is a fairly thin client. I'll raise internally whether this can be prioritized.

**Comment by quartzmo** [comment]
@tritone Can you please ask internally about the similar request to add this in google-cloud-ruby as well?

googleapis/google-cloud-ruby#8147

**Comment by quartzmo** [comment]
@tritone Any update as to whether or not this feature can be prioritized?

**Comment by tritone** [comment]
@quartzmo I don't think we are prioritizing this in the next quarter at least. @frankyn may have more context for Ruby specifically.

**Comment by frankyn** [comment]
We are not yet prioritizing this work across languages. It's still in our backlog, and we are working our way back to it.

**Comment by Slach** [comment]
@frankyn why you don't want implement it?

**Comment by frankyn** [comment]
We do, but it's competing with other work items don't have ETA yet. Thanks for your patience.

**Comment by tritone** [comment]
Just wanted to provide a long-delayed update here. We're planning on implementing parallel upload and download features in the newly-created subpackage transfermanager. We did the initial merge for downloads this week, more download features are coming later this month, and uploads should be coming in the second half of the year most likely. Feel free to take a look here: https://pkg.go.dev/cloud.google.com/go/storage@v1.41.1-0.20240605002059-de9872df8565/transfermanager ; we are open to hearing feedback.

**Comment by Slach** [comment]
@tritone great news, hope you will update this issue when Uploader will implements

**Comment by Vytenis-Valutkevicius** [comment]
Hi, @tritone, I am rather interested in this feature, so since it's been few months since last update just wanted to as, maybe you know if it is currently scheduled or some ETA is assigned? Thanks.

**Comment by begelundmuller** [comment]
Hi @tritone, we're also very interested in Go support for this feature! Just wanted to ask when you think it might be ready for production use? Thanks!

**Comment by pingleiwandb** [comment]
Base on https://pkg.go.dev/cloud.google.com/go/storage/transfermanager and https://github.com/googleapis/google-cloud-go/tree/storage/v1.51.0/storage/transfermanager I guess it is already released and ready for use?

**Comment by cpriti-os** [comment]
Update: We have picked up the work to support parallel composite uploads and should be available by the end of this quarter. Parallel uploads within the subpackage transfer manager will follow up soon.

**Comment by vis21** [comment]
hey @cpriti-os, what's the status of this issue if you don't mind me asking? We're close to the end of the quarter and it'd be good to know what to expect.

**Comment by cpriti-os** [comment]
Hi, the parallel composite uploads feature is almost developed, support for Parallel transfers within the Writer surface will be out in the next release. (1-2 weeks).

**Comment by cpriti-os** [comment]
It's released now with experimental tag in v1.62, will move it to stable in a few weeks after monitoring.
